src/stream/src/executor/dynamic_filter.rs

```diff
@@ -0,0 +1,339 @@
+
+use std::collections::BTreeMap;
+use std::ops::Bound::*;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use itertools::Itertools;
+use madsim::collections::HashSet;
+use risingwave_common::array::{Op, Row, StreamChunk};
+use risingwave_common::buffer::BitmapBuilder;
+use risingwave_common::catalog::Schema;
+use risingwave_common::types::{Datum, ScalarImpl, ToOwnedDatum};
+use risingwave_expr::expr::BoxedExpression;
+use risingwave_pb::expr::expr_node::Type as ExprNodeType;
+use risingwave_pb::expr::expr_node::Type::*;
+
+use super::barrier_align::*;
+use super::error::StreamExecutorError;
+use super::monitor::StreamingMetrics;
+use super::{BoxedExecutor, BoxedMessageStream, Executor, Message, PkIndices, PkIndicesRef};
+use crate::common::StreamChunkBuilder;
+use crate::executor::PROCESSING_WINDOW_SIZE;
+
+pub struct DynamicFilterExecutor {
+    source_l: BoxedExecutor,
+    source_r: BoxedExecutor,
+    key_l: usize,
+    pk_indices: PkIndices,
+    identity: String,
+    cond: BoxedExpression,
+    comparator: ExprNodeType,
+    actor_id: u64,
+    schema: Schema,
+    metrics: Arc<StreamingMetrics>,
+}
+
+impl DynamicFilterExecutor {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        source_l: BoxedExecutor,
+        source_r: BoxedExecutor,
+        key_l: usize,
+        pk_indices: PkIndices,
+        executor_id: u64,
+        cond: BoxedExpression,
+        comparator: ExprNodeType,
+        actor_id: u64,
+        metrics: Arc<StreamingMetrics>,
+    ) -> Self {
+        let schema = source_l.schema().clone();
+        Self {
+            source_l,
+            source_r,
+            key_l,
+            pk_indices,
+            identity: format!("DynamicFilterExecutor {:X}", executor_id),
+            cond,
+            comparator,
+            actor_id,
+            metrics,
+            schema,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn into_stream(self) {
+        let mut prev_epoch_value: Option<Datum> = None;
+        let mut current_epoch_value: Option<Datum> = None;
+
+        // The state is sorted by the comparison value
+        //
+        // TODO: convert this into a `StateTable` compatible managed state
+        //
+        // TODO: It could be potentially expensive memory-wise to store `HashSet`.
+        //       If I'm not wrong, the memory overhead is a backing Vec of size 4
+        //       (See: https://github.com/rust-lang/hashbrown/pull/162)
+        //       + some byte-per-entry metadata. Well, `Row` is on heap anyway...
+        //       It could be preferred to find a way to do prefix range scans on the left key and
+        //       storing as `BTreeSet<(ScalarImpl, Row)>`.
+        //       We could solve it if `ScalarImpl` had a successor/predecessor function.
+        //
+        //       Probably, using a custom comparator function on a custom datatype
```

Comment:
Actually, I will do this in this PR

Reply:
Basically, what we want is a multimap. But there is currently no efficient implementation for such a multimap.